### PR TITLE
[refactor] Remove Program::KernelProxy

### LINKED ISSUE #2723: Remove Program::KernelProxy
`KernelProxy` is defined around:

https://github.com/taichi-dev/taichi/blob/63140f6fa56322c5af606662a60f48cb50acea2b/taichi/program/program.h#L194-L202

It complicates the readability without adding much value

https://github.com/taichi-dev/taichi/blob/63140f6fa56322c5af606662a60f48cb50acea2b/python/taichi/lang/kernel_impl.py#L434-L452

* L434: Creates a `KernelProxy` using https://github.com/taichi-dev/taichi/blob/63140f6fa56322c5af606662a60f48cb50acea2b/taichi/python/export_lang.cpp#L762-L765
* L452: Creates an actual `Kernel` using https://github.com/taichi-dev/taichi/blob/63140f6fa56322c5af606662a60f48cb50acea2b/taichi/program/program.h#L199-L201

**Describe the solution you'd like (if any)**

Just pass `taichi_ast_generator` to `Kernel`'s constructor directly. We don't have to use KernelProxy.

### PR DESCRIPTION
**Pull request by lin-hitonami**
Related issue = close #2723

### CODE REVIEW

File: taichi/python/export_lang.cpp [hunk 1]
```diff
@@ -772,10 +763,13 @@ void export_lang(py::module &m) {
   m.def("get_external_tensor_shape_along_axis",
         Expr::make<ExternalTensorShapeAlongAxisExpression, const Expr &, int>);
 
-  m.def("create_kernel",
-        [&](std::string name, bool grad) -> Program::KernelProxy {
-          return get_current_program().kernel(name, grad);
-        });
+  m.def(
+      "create_kernel",
+      [&](const std::function<void()> &body, const std::string &name = "",
```
Comment (ailzhang) [inline taichi/python/export_lang.cpp:768]: nit: kernel_name was a required argument before, is there any reason that we want to make it optional here? ;)

**Review by ailzhang — APPROVED** [review]
Overall looks great! Thanks!

File: taichi/python/export_lang.cpp (hunk 1 above)
Comment (lin-hitonami) [reply taichi/python/export_lang.cpp]: fixed

File: taichi/python/export_lang.cpp [hunk 2]
```diff
@@ -424,15 +424,6 @@ void export_lang(py::module &m) {
       .def("serialize", &ExprGroup::serialize);
 
   py::class_<Stmt>(m, "Stmt");
-  py::class_<Program::KernelProxy>(m, "KernelProxy")
-      .def(
-          "define",
-          [](Program::KernelProxy *ker,
-             const std::function<void()> &func) -> Kernel * {
-            py::gil_scoped_release release;
```
Comment (ailzhang) [inline taichi/python/export_lang.cpp:432]: @lin-hitonami This `gil_scoped_release` was accidentally removed in the refactor? Would you mind adding it back? 
Thanks @strongoier for catching this!

**Pull request merged**